### PR TITLE
document random reproducibility policy

### PR DESCRIPTION
**Pull request by stevengj**
See e.g. discussion in #30494 and [on discourse](https://discourse.julialang.org/t/dealing-with-seeds-in-random-numbers/29029).

See also https://numpy.org/neps/nep-0019-rng-policy.html for a similar discussion by the NumPy developers.

### CODE REVIEW

**Comment by tpapp** [comment]
Thanks for writing this up.

Do you think we can/should make a commitment to reproducibility across OSs/architectures? The native Julia code should have this property (though we should be careful about endianness if the issue arises), but I don't know if the dSMFT code can guarantee this.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by stevengj** [comment]
I would hope that the MT code would guarantee portability, since it is supposed to be following a specific number-theoretic sequence...

**Comment by stevengj** [comment]
Should be good to merge?   CI failure (`cp: cannot stat 'dist-extras/7z.*': No such file or directory`) is unrelated, obviously.

**Comment by KristofferC** [comment]
cc @staticfloat for the win failure

**Comment by GregPlowman** [comment]
> I would hope that the MT code would guarantee portability, since it is supposed to be following a specific number-theoretic sequence...

Yes, but perhaps higher-level rand functions might be architecture-specific.
32-bit vs 64 bit
Alternative/future RNGs (other than MT)
Secure/crypto rands with hardware support

Would it be better to explicitly state policy w.r.t architecture? 


https://github.com/JuliaLang/julia/pull/29240#issuecomment-497363751
https://github.com/JuliaLang/julia/pull/29240#issuecomment-498013285
https://github.com/JuliaLang/julia/pull/29240#issuecomment-498019043

File: stdlib/Random/docs/src/index.md [hunk 1]
```diff
@@ -326,3 +326,14 @@ The non-mutating array method of `rand` will automatically call this specializat
...
+
+If you need to guarantee exact reproducibility of random data, it is advisable to simply *save the data* (e.g. as a supplementary attachment in a scientific publication).  (You can also, of course, specify a
+particular Julia version and package manifest, especially if you require bit reproducibility.)
+
+Software tests that rely on *specific* "random" data should also generally save the data or embed it into the test code.  On the other hand, tests that should pass for *most* random data (e.g. testing `A \ (A*x) ≈ x` for a random matrix `A = randn(n,n)`) can use an RNG with a fixed seed to ensure that simply running the test many times does not encounter a failure due to very improbable data (e.g. an extremely ill-conditioned matrix).
+
+The statistical *distribution* from which random samples are drawn *is* guaranteed to be the same across any minor Julia releases.
```
Comment (rfourquet) [inline stdlib/Random/docs/src/index.md]: Would this rule out the possibility to change the distribution of `rand(::Type{<:AbstractFloat})` from uniform in `[0,1)` into uniform in `(0,1)` which is currently discussed? or is this a "minor" enough change?

Comment (StefanKarpinski) [reply stdlib/Random/docs/src/index.md]: In measure theoretic terms, those are technically the same distribution 😬

**Pull request merged**